docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,130 @@
+Nullable Metadata
+=========
+The following describes the representation of nullable annotations in metadata.
+
+### NullableAttribute
+Type references are annotated in metadata with a `NullableAttribute`.
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(
+        AttributeTargets.Class |
+        AttributeTargets.Event |
+        AttributeTargets.Field |
+        AttributeTargets.GenericParameter |
+        AttributeTargets.Parameter |
+        AttributeTargets.Property |
+        AttributeTargets.ReturnValue,
+        AllowMultiple = false)]
+    public sealed class NullableAttribute : Attribute
+    {
+        public readonly byte[] NullableFlags;
+        public NullableAttribute(byte b)
+        {
+            NullableFlags = new byte[] { b };
+        }
+        public NullableAttribute(byte[] b)
+        {
+            NullableFlags = b;
+        }
+    }
+}
+```
+
+Each type reference is accompanied by a `NullableAttribute` with an array of bytes, where 0 is oblivious, 1 is not annotated and 2 is annotated.
+
+Value types may be marked with 0 (oblivious) or 1 (not annotated) - the representations are considered equivalent.
+
+To optimize trivial cases the attribute may be omitted or replaced with an attribute that takes a single byte value rather than an array. The C#8 compiler uses the following optimizations.
+
+Optimized cases:
+1. All parts are oblivious in oblivious [emit context](#NullableContextAttribute): the `NullableAttribute` is omitted
+2. All parts are oblivious in not annotated [emit context](#NullableContextAttribute): `NullableAttribute(0)` (rather than `new byte[] { 0 }`)
+3. All parts are not annotated in not annotated [emit context](#NullableContextAttribute): the `NullableAttribute` is omitted
+4. All parts are not annotated in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)` (rather than `new byte[] { 1 }`)
+5. All parts are annotated: `NullableAttribute(2)` (rather than `new byte[] { 2 }`)
+
+The first case, where all parts are oblivious in an oblivious context, matches the representation in legacy assemblies where there are no `NullaleAttribute` annotations.
+
+The following `NullableAttribute` cases are recognized and emitted for type parameter definitions:
+1. `notnull` constraint in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)`
+2. `class` constraint in `#nullable enable` and in oblivious [emit context](#NullableContextAttribute): `NullableAttribute(1)`
+2. `class` constraint in `#nullable disable` and in not annotated [emit context](#NullableContextAttribute): `NullableAttribute(0)`
+3. `class?` constraint: `NullableAttribute(2)`
+4. No `notnull`, `class`, `struct`, `unmanaged`, or type constraints in `#nullable enable` (the equivalent of an `object?` constraint): `NullableAttribute(2)`
```

Comment:
In 'not annotated' context:
`notnull` - No attribute
No constraints in `#nullable enable` - `NullableAttribute(2)`
No constraints in `#nullable disable` - If we do not emit anything how do we distinguish this case from `notnull`?

It might be good to enumerate all scenarios per NullableContextAttribute (rather than having items corresponding to different contexts in one list) to make sure there is no conflict/ambiguity.